### ISSUE TITLE
Generic React HOC: `TypeError: Cannot read property 'declarations' of undefined`

### ISSUE THREAD

**Issue by mingcenwei** [microsoft/TypeScript#41616]
**TypeScript Version:**  4.1.2

**Search Terms:** TypeError Cannot read property declarations of undefined

**Code**
`src-file1.tsx`:
```ts
import React from "react";

type Tags = "span" | "div";

export const Hoc = <Tag extends Tags>(
	TagElement: Tag,
): React.FunctionComponent => {
	const Component = () => <TagElement />;
	return Component;
};
```
`src-file2.tsx`:
```ts
import React from "react";

type Tags = "span" | "div";

export const Hoc = (TagElement: Tags): React.FunctionComponent => {
	const Component = () => <TagElement />;
	return Component;
};
```

**Expected behavior:**
Both `npx tsc --strict --noEmit --jsx 'react' src-file1.tsx` and `npx tsc --strict --noEmit --jsx 'react' src-file2.tsx` will run successfully.

**Actual behavior:**
`npx tsc --strict --noEmit --jsx 'react' src-file2.tsx` runs successfully, but `npx tsc --strict --noEmit --jsx 'react' src-file1.tsx` fails with the error message (where `"$PWD"` is the present working directory and `"$HOME"` is my home directory):
```
"$PWD"/node_modules/typescript/lib/tsc.js:87001
                throw e;
                ^

TypeError: Cannot read property 'declarations' of undefined
    at addImplementationSuccessElaboration ("$PWD"/node_modules/typescript/lib/tsc.js:57022:116)
    at resolveCall ("$PWD"/node_modules/typescript/lib/tsc.js:56947:33)
    at resolveJsxOpeningLikeElement ("$PWD"/node_modules/typescript/lib/tsc.js:57576:20)
    at resolveSignature ("$PWD"/node_modules/typescript/lib/tsc.js:57597:28)
    at getResolvedSignature ("$PWD"/node_modules/typescript/lib/tsc.js:57608:26)
    at checkJsxOpeningLikeElementOrOpeningFragment ("$PWD"/node_modules/typescript/lib/tsc.js:55477:27)
    at checkJsxSelfClosingElementDeferred ("$PWD"/node_modules/typescript/lib/tsc.js:55058:13)
    at checkDeferredNode ("$PWD"/node_modules/typescript/lib/tsc.js:64402:21)
    at Map.forEach (<anonymous>)
    at checkDeferredNodes ("$PWD"/node_modules/typescript/lib/tsc.js:64373:37)
npm ERR! code 1
npm ERR! path "$PWD"
npm ERR! command failed
npm ERR! command sh -c tsc --strict --noEmit --jsx react --esModuleInterop src-file1.tsx

npm ERR! A complete log of this run can be found in:
npm ERR!    "$HOME"/.npm/_logs/2020-11-20T20_34_21_223Z-debug.log
```

**Playground Link:** 
Compiling [the first source file](https://www.typescriptlang.org/play?ts=4.2.0-dev.20201120#code/JYWwDg9gTgLgBAJQKYEMDG8BmUIjgIilQ3wG4AocmATzCTgBUUBzAZzgF4DWwUA7fHAA+BACbAAbmUpoIfVvAASENJzgAeJszhIAHjCR9R7LawB8ACnIBILQFEANkhCGYALkYsANOQCUH5HQYADoAMQBXPgxgOQBhXEg+V04zOABvG1l5eHjwOWSuC18UjXsnFz54AHozCmsiGHCoPjhcxNcKAF9SIA) crashes the web compiler, 
<img width="1437" alt="image" src="https://user-images.githubusercontent.com/47605337/99848688-300a7a00-2bb5-11eb-9946-6cdd4dbcdd3c.png">
<img width="1439" alt="image" src="https://user-images.githubusercontent.com/47605337/99848710-37ca1e80-2bb5-11eb-8b99-28f1df0a1739.png">
<img width="1437" alt="image" src="https://user-images.githubusercontent.com/47605337/99848249-65fb2e80-2bb4-11eb-92c6-23df11816a20.png">
while compiling [the second source file](https://www.typescriptlang.org/play?ts=4.2.0-dev.20201120#code/JYWwDg9gTgLgBAJQKYEMDG8BmUIjgIilQ3wG4AocmATzCTgBUUBzAZzgF4DWwUA7fHAA+BACbAAbmUpoIfVvAASENJzgAKJswCiAGyQgkfGAC5GLVgEozydDAB0AMQCufDMDkBhXJD5H4HAB8cADe5ACQsvLw3uBy-mrqlpzBADxaegYJAPSBFOFEMM5QfHCxvv4UAL6kQA) succeeds.
<img width="1437" alt="image" src="https://user-images.githubusercontent.com/47605337/99848740-44e70d80-2bb5-11eb-971b-b5e4f4c1bcf0.png">
<img width="1437" alt="image" src="https://user-images.githubusercontent.com/47605337/99848759-4b758500-2bb5-11eb-9dca-3cd5f88d2958.png">
<img width="1438" alt="image" src="https://user-images.githubusercontent.com/47605337/99848543-ee79cf00-2bb4-11eb-805e-bbc5b37a132e.png">

**Related Issues:** 
#33735 #21817 #21811 #21798 #20475 #6428

**Comment by mingcenwei** [comment]
Using `React.createElement(TagElement)` instead of `<TagElement />` is a workaround.

**Comment by capaj** [comment]
we also get this error on our production React app at @leaplabs

**Comment by remcohaszing** [comment]
We’re also getting this error [here](https://gitlab.com/appsemble/appsemble/-/jobs/871257539). This doesn’t use React, but it does use JSX. The relevant code can be found [here](https://gitlab.com/appsemble/appsemble/-/tree/0.15.9/blocks/button-list).

**Comment by lkarmelo** [comment]
I got the same issue in simple Button component code, even if i simplify it to this
```javascript
const Element = 'button';
return (
   <Element
     //some props
    />
)
```

Returning just ```<button />``` works fine

**Comment by olee** [comment]
Could someone perhaps tell me in which version this was introduced so I can downgrade typescript again as minimal as possible?

**Comment by remcohaszing** [comment]
This bug was introduced in version 4.1.0 and still exists in 4.1.2  
The last working version is 4.0.5  
A fix was made in #41758 and will be released in 4.2.0  
A merge request to include this in 4.1.x is pending in #41862

Until #41862 is released, people should stick to version 4.0.5.
